aten/src/ATen/cuda/CUDABlas.cpp

```diff
@@ -0,0 +1,325 @@
+    cublasStatus_t __err = EXPR;          \
+    if (__err != CUBLAS_STATUS_SUCCESS) { \
+      AT_ERROR(                           \
+          "CUDA error: ",                 \
+          _cublasGetErrorEnum(__err),     \
+          " when calling `" #EXPR "`");   \
+    }                                     \
+  } while (0)
+
+#define CUDABLAS_POSINT_CHECK(FD, X)         \
+  TORCH_CHECK(                               \
+      (X > 0 && X <= INT_MAX),               \
+      "at::cuda::blas::" #FD " argument " #X \
+      " must be positive and less than ",    \
+      INT_MAX,                               \
+      " but got ",                           \
+      X)
+
+#define CUDABLAS_NONNEGINT_CHECK(FD, X)       \
+  TORCH_CHECK(                                \
+      (X >= 0 && X <= INT_MAX),               \
+      "at::cuda::blas::" #FD " argument " #X  \
+      " must be non-negative and less than ", \
+      INT_MAX,                                \
+      " but got ",                            \
+      X)
+
+namespace {
+
+static const char* _cublasGetErrorEnum(cublasStatus_t error) {
+  if (error == CUBLAS_STATUS_SUCCESS) {
+    return "CUBLAS_STATUS_SUCCESS";
+  }
+  if (error == CUBLAS_STATUS_NOT_INITIALIZED) {
+    return "CUBLAS_STATUS_NOT_INITIALIZED";
+  }
+  if (error == CUBLAS_STATUS_ALLOC_FAILED) {
+    return "CUBLAS_STATUS_ALLOC_FAILED";
+  }
+  if (error == CUBLAS_STATUS_INVALID_VALUE) {
+    return "CUBLAS_STATUS_INVALID_VALUE";
+  }
+  if (error == CUBLAS_STATUS_ARCH_MISMATCH) {
+    return "CUBLAS_STATUS_ARCH_MISMATCH";
+  }
+  if (error == CUBLAS_STATUS_MAPPING_ERROR) {
+    return "CUBLAS_STATUS_MAPPING_ERROR";
+  }
+  if (error == CUBLAS_STATUS_EXECUTION_FAILED) {
+    return "CUBLAS_STATUS_EXECUTION_FAILED";
+  }
+  if (error == CUBLAS_STATUS_INTERNAL_ERROR) {
+    return "CUBLAS_STATUS_INTERNAL_ERROR";
+  }
+#if CUDA_VERSION >= 6000
+  if (error == CUBLAS_STATUS_NOT_SUPPORTED) {
+    return "CUBLAS_STATUS_NOT_SUPPORTED";
+  }
+#endif
+#if CUDA_VERSION >= 6050
+  if (error == CUBLAS_STATUS_LICENSE_ERROR) {
+    return "CUBLAS_STATUS_LICENSE_ERROR";
+  }
+#endif
+  return "<unknown>";
+}
+
+static cublasOperation_t _cublasOpFromChar(char op) {
+  switch (op) {
+    case 'n':
+    case 'N':
+      return CUBLAS_OP_N;
+    case 't':
+    case 'T':
+      return CUBLAS_OP_T;
+    case 'c':
+    case 'C':
+      return CUBLAS_OP_C;
+  }
+  AT_ERROR(
+      "_cublasOpFromChar input should be 't', 'n' or 'c' but got `", op, "`");
+}
+
+static void _cublasAdjustLdLevel2(int64_t m, int64_t n, int64_t* lda) {
+  // Note: leading dimensions generally are checked that they are > 0
+  // and at least as big the result requires (even if the value won't
+  // be used).
+
+  // Q: Why does Level3 check trans but this doesn't?
+  // A: In level 2, the sizes (m, n) specify the size of A
+  // (independent of trans value). In level 3. the sizes (m, n, k)
+  // specify the sizes of op(A), op(B) where op depend on trans
+  // values.
```

Comment:
Cool, thank you for figuring this out :)

Reply:
no problem:)